src/native/managed/cdac/mscordaccore_universal/Legacy/ClrDataFrame.cs

```diff
@@ -53,7 +107,70 @@ int IXCLRDataFrame.GetArgumentByIndex(
         => _legacyImpl is not null ? _legacyImpl.GetArgumentByIndex(index, arg, bufLen, nameLen, name) : HResults.E_NOTIMPL;
 
     int IXCLRDataFrame.GetNumLocalVariables(uint* numLocals)
-        => _legacyImpl is not null ? _legacyImpl.GetNumLocalVariables(numLocals) : HResults.E_NOTIMPL;
+    {
+        int hr = HResults.S_OK;
+        try
+        {
+            IStackWalk stackWalk = _target.Contracts.StackWalk;
+            TargetPointer methodDesc = stackWalk.GetMethodDescPtr(_dataFrame);
+            if (methodDesc == TargetPointer.Null)
+                throw Marshal.GetExceptionForHR(/*E_NOINTERFACE*/ HResults.COR_E_INVALIDCAST)!;
+
+            IRuntimeTypeSystem rts = _target.Contracts.RuntimeTypeSystem;
+            MethodDescHandle mdh = rts.GetMethodDescHandle(methodDesc);
+            if (!rts.IsIL(mdh))
+            {
+                *numLocals = 0;
+                throw Marshal.GetExceptionForHR(HResults.E_FAIL)!;
+            }
+
+            // get token and module from method desc
+            TargetPointer mtAddr = rts.GetMethodTable(mdh);
+            TypeHandle typeHandle = rts.GetTypeHandle(mtAddr);
+            TargetPointer modulePtr = rts.GetModule(typeHandle);
+            ILoader loader = _target.Contracts.Loader;
+            Contracts.ModuleHandle moduleHandle = loader.GetModuleHandleFromModulePtr(modulePtr);
+            uint token = rts.GetMethodToken(mdh);
+
+            TargetPointer ilHeader = loader.GetILHeader(moduleHandle, token);
+            int localToken = HeaderReaderHelpers.GetLocalVarSigToken(_target, ilHeader);
+            if (localToken < 0)
+            {
+                *numLocals = 0;
+                throw Marshal.GetExceptionForHR(HResults.E_FAIL)!;
+            }
+
+            IEcmaMetadata ecmaMetadataContract = _target.Contracts.EcmaMetadata;
+            MetadataReader? mdReader = ecmaMetadataContract.GetMetadata(moduleHandle);
+            if (mdReader == null)
+                throw new NotImplementedException();
+
+            StandaloneSignatureHandle localSignatureHandle = MetadataTokens.StandaloneSignatureHandle(localToken);
+            BlobHandle localSignatureBlob = mdReader.GetStandaloneSignature(localSignatureHandle).Signature;
+            BlobReader blobReader = mdReader.GetBlobReader(localSignatureBlob);
+
+            // see ECMA-335 II.23.2.6
+            SignatureHeader header = blobReader.ReadSignatureHeader();
+            if (header.Kind != SignatureKind.LocalVariables)
+                throw new BadImageFormatException();
+            *numLocals = (uint)blobReader.ReadCompressedInteger();
+        }
+        catch (System.Exception ex)
+        {
+            hr = ex.HResult;
```

Comment:
```suggestion
            *numLocals = 0;
            hr = ex.HResult;
```
Clear numLocals here (or at at start of the method) so that it is initialized in all error cases?